src/wallet/coinselection.cpp

```diff
@@ -0,0 +1,300 @@
+            if (curr_waste <= best_waste) {
+                best_selection.assign(selection.begin(), selection.end());
+                best_waste = curr_waste;
+            }
+            curr_waste -= (value_ret - target_value); // Remove the excess value as we will be selecting different coins now
+            backtrack = true;
+        } else if (depth >= (int)utxo_pool.size()) { // Reached a leaf node, no solution here
+            backtrack = true;
+        } else if (value_ret + lookahead < target_value) { // Cannot possibly reach target with the amount remaining in the lookahead
+            if (depth == 0) { // At the first utxo, no possible selections, so exit
+                break;
+            } else {
+                backtrack = true;
+            }
+        } else { // Continue down this branch
+            // Assert that this utxo is not negative. It should never be negative, effective value calculation should have removed it
+            assert(utxo_pool.at(depth).txout.nValue >= 0);
+
+            // Remove this utxo from the lookahead utxo amount
+            lookahead -= utxo_pool.at(depth).txout.nValue;
+            // Increase waste
+            curr_waste += (utxo_pool.at(depth).fee - utxo_pool.at(depth).long_term_fee);
+            // Inclusion branch first (Largest First Exploration)
+            selection.at(depth).first = true;
+            value_ret += utxo_pool.at(depth).txout.nValue;
+            ++depth;
+        }
+
+        // Step back to the previous utxo and try the other branch
+        if (backtrack) {
+            backtrack = false; // Reset
+            --depth;
+
+            // Walk backwards to find the first utxo which has not has its second branch traversed
+            while (selection.at(depth).second) {
+                selection.at(depth).first = false;
+                selection.at(depth).second = false;
+                lookahead += utxo_pool.at(depth).txout.nValue;
+
+                // Step back one
+                --depth;
+
+                if (depth < 0) { // We have walked back to the first utxo and no branch is untraversed. No solution, exit.
+                    done = true;
+                    break;
+                }
+            }
+
+            if (!done) {
+                // Now traverse the second branch of the utxo we have arrived at.
+                selection.at(depth).second = true;
+
+                // These were always included first, try excluding now
+                selection.at(depth).first = false;
+                value_ret -= utxo_pool.at(depth).txout.nValue;
+                curr_waste -= (utxo_pool.at(depth).fee - utxo_pool.at(depth).long_term_fee);
+                ++depth;
+            }
+        }
+        --remaining_tries;
+    }
+
+    // Check for solution
+    if (best_selection.empty()) {
+        return false;
+    }
+
+    // Set output set
+    value_ret = 0;
```

Comment:
Why are you recalculating value_ret, when you've been tracking it above before?

Reply:
The `value_ret` above is not necessarily our best set. It is used for tracking our work during the selection so it may not actually be the value for the selection we want to use. It should probably be a different variable.

Reply:
Yeah, let's make it two different variables with clear names.